### ISSUE TITLE
HuBERT feature size error during concatenation

### ISSUE THREAD

**Issue by ShaunBuckley** [RF5/simple-asgan#1]
Hi there, I've been trying to train your model with HuBERT features extracted from the sc09 dataset, but when I run the train_asgan.py file using the recommended config, I get this error:

Sizes of tensors must match except in dimension 0. Expected size 50 but got size 43 for tensor number 37 in the list

This occurs on the line: 

data = torch.cat([self.__getitem__(i)[0] for i in progress_bar(range(len(self)))], dim=0).float()

contained in the prep_fad_metrics() function within the HubertFeatureDataset() class.

I am extracting the HuBERT features using your recommended process:

  import torch
  import torchaudio

  hubert = torch.hub.load('RF5/simple-asgan', 'hubert_base')
  wav, sr = torchaudio.load('/path/to/google_speech_commands_utterance.wav')

  feats = hubert.get_feats_batched(wav) # (bs, seq_len, dim)
  torch.save(feats, '/path/to/hubert_feature_cache/uttr_hubert_feats.pt')

contained within a loop through all of the files, saving the files into a directory structure and naming convention identical to sc09 with .pt suffixes.

This error seems to stem from the differences in sequence length of the samples, so I was wondering if you knew how to deal with this error or could potentially share the code you used to preprocess the sc09 dataset into HuBERT features for training.

**Comment by RF5** [comment]
Hi @ShaunBuckley , thanks for your question.

One big limitation of the ASGAN model architecture is that it requires all the inputs to be of equal length. For SC09 it is not an issue because they are all roughly a second long. There were a few samples in SC09 that were just under a second long, and those I padded to be exactly a second long. 

So, in short, to use your own dataset you will need to pad your features or input audios to equal length before training the GAN. Alternatively (and a bit more effort) would be to find a way to modify ASGAN to accept variable-length inputs. It is, however, a known current limitation that it does not work nicely with longer inputs, so that may be a bit more effort.

I hope that helps, and please let me know if you get it to work on your dataset!

**Comment by ShaunBuckley** [comment]
Hi @RF5, thanks for your quick reply!

Yes I figured there may need to be padding, I tried this but unfortunately ran into another error in the collate function of the dataset file on this line:

gt_feats[i, :l] = xs[i][0][:l]

The error is:

The expanded size of the tensor (1) must match the existing size (50) at non-singleton dimension 0.  Target sizes: [1, 768].  Tensor sizes: [50, 768]



The dataset I'm using is the same as your paper, sc09, and I padded each waveform to 1 second using the following code:

data, sr = librosa.load(path, sr=16000)
durationSeconds = 1
padded = librosa.util.fix_length(data=data, size=int(sr * durationSeconds))
wav = torch.from_numpy(padded).unsqueeze(0)

This wav is then passed to the HuBERT feature extractor, giving shape [1, 50, 768] as intended.

The run config I'm passing as arguments (the config file is left as default) is as follows:

model=rp_w train_root=hubert_feature_cache_padded/ n_valid=400 data_type=hubert_L6 checkpoint_path=./density/runs/cool_training_run/ z_dim=512 rp_w_cfg.z_dim=512 rp_w_cfg.w_layers=1 batch_size=16 lr=2e-3 grad_clip=10 aug_init_p=0.2 stdout_interval=100 validation_interval=2500 n_epochs=800 c_dim=768 rp_w_cfg.c_dim=768 d_lr_mult=0.1 fp16=True preload=False num_workers=12 betas=[0,0.99] rp_w_cfg.equalized_lr=True rp_w_cfg.use_sg3_ff=True rp_w_cfg.D_kernel_size=5 rp_w_cfg.D_block_repeats=[3,3,3,3] use_sc09_splits=True sc09_train_csv=./splits/sc09-train.csv sc09_valid_csv=./splits/sc09-valid.csv rp_w_cfg.r1_gamma=0.1


If there is any more info that could be helpful for debugging please just let me know, and thanks again for your help!

**Comment by RF5** [comment]
Hmm, it looks like your `gt_feats` in preprocessing has an extra dimension from the logs. In other words, `gt_feats` should be a tensor of shape `(batch_size, 50, 768)`, but it looks like it is shape `(batch_size, 1, 50, 768)` for some reason. What batch size are you using? Could it be that the silence vector you are using has an extra dimension maybe? Or possibly were the features saved with an additional dimension? The dataloader [here](https://github.com/RF5/simple-asgan/blob/master/density/dataset.py#L73) assumes the HuBERT features to be (seq_len, 768) and not (1, seq_len, 768) -- could that be it? 

I hope one of those leads helps a bit. Luckily it looks like only tensor shape/formatting problem. Let me know how it goes!

**Comment by ShaunBuckley** [comment]
Based on what you've said, I had a look at the preprocessing and it turns out the hubert.get_feats_batched() function in the block of code provided as an example in the readme for saving a wav file was actually producing a tensor of [1, seq_len, 768]. Adding a simple squeeze(0) line to the feats before saving as a .pt file seems to have fixed these issues. Thanks again for your help!

**Comment by RF5** [comment]
Ahh thank you for the catch, I will update the readme to add a note about that! Glad it works now :)
